lib/Doctrine/DBAL/Connections/PrimaryReplicaConnection.php

```diff
@@ -0,0 +1,361 @@
+    }
+
+    /**
+     * @param mixed[] $params
+     *
+     * @return mixed[]
+     */
+    private function chooseConnectionConfiguration(string $connectionName, array $params): array
+    {
+        if ($connectionName === 'primary') {
+            return $params['primary'];
+        }
+
+        $config = $params['replica'][array_rand($params['replica'])];
+
+        if (! isset($config['charset']) && isset($params['primary']['charset'])) {
+            $config['charset'] = $params['primary']['charset'];
+        }
+
+        return $config;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function executeUpdate($query, array $params = [], array $types = []): int
+    {
+        $this->switchToPrimary();
+
+        return parent::executeUpdate($query, $params, $types);
+    }
+
+    public function beginTransaction(): bool
+    {
+        $this->switchToPrimary();
+
+        return parent::beginTransaction();
+    }
+
+    public function commit(): bool
+    {
+        $this->switchToPrimary();
+
+        return parent::commit();
+    }
+
+    public function rollBack(): bool
+    {
+        $this->switchToPrimary();
+
+        return parent::rollBack();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function delete($tableName, array $identifier, array $types = []): int
+    {
+        $this->switchToPrimary();
+
+        return parent::delete($tableName, $identifier, $types);
+    }
+
+    public function close(): void
+    {
+        unset($this->primary, $this->replica);
+
+        parent::close();
+
+        $this->_conn = null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function update($tableName, array $data, array $identifier, array $types = []): int
+    {
+        $this->switchToPrimary();
+
+        return parent::update($tableName, $data, $identifier, $types);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function insert($tableName, array $data, array $types = []): int
+    {
+        $this->switchToPrimary();
+
+        return parent::insert($tableName, $data, $types);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function exec($statement): int
+    {
+        $this->switchToPrimary();
+
+        return parent::exec($statement);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function createSavepoint($savepoint): void
+    {
+        $this->switchToPrimary();
+
+        parent::createSavepoint($savepoint);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function releaseSavepoint($savepoint): void
+    {
+        $this->switchToPrimary();
+
+        parent::releaseSavepoint($savepoint);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function rollbackSavepoint($savepoint): void
+    {
+        $this->switchToPrimary();
+
+        parent::rollbackSavepoint($savepoint);
+    }
+
+    public function query(): Statement
+    {
+        $this->switchToPrimary();
```

Comment:
And it's even worse here. Why are we switching to the primary endpoint to execute a _query_?

Reply:
I won't like, I have no idea…

Reply:
the query function can be used for any kind of SQL, including write operations. This cannot be detected without parsing the SQL statement, which is a rabbit hole. This is why for safety reason the connectoin assumes that when using $connection->query(), you might be performing a write operation.

Reply:
This is why the developer should be in control of which connection to use for a given statement. Not the abstraction layer.